internal/search/parser.go

```diff
@@ -0,0 +1,283 @@
+package search
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+)
+
+type Node interface {
+	String() string
```

Comment:
You could add two methods here:
```suggestion
	String() string
	IsParameter() bool
	IsOp() bool
```

and implement them on `Parameter` and `Op`.

That would allow you to get rid of some of the type switches/checks further down and it would make it harder to accidentally implement one of these interfaces. Example: everything that implements [fmt.Stringer](https://golang.org/pkg/fmt/#Stringer) could slip through as a `Node`. I don't think it's a real danger, but just pointing it out there, that there _is_ a practice of giving some Go interfaces private methods that are only implemented by types in the package, so that user-defined types won't slip into the package boundaries.

Reply:
> there is a practice of giving some Go interfaces private methods that are only implemented by types in the package, so that user-defined types won't slip into the package boundaries.

This makes sense to me. For all practical purposes, we could define a unique identifier like `haha_im_a_node_cd5faa62` as a private function member to avoid the unwanted behavior here. We should choose something more reasonable, but the point is that in this model the function should not do anything meaningful. I'm in favor of that. That said, having `IsParameter`/`IsOp` functions, in contrast, would be _meaningful_ and multiplicative (define `isX` for every term in the language). As @keegancsmith mentioned, these are mutually exclusive (by definition of being terms) and I just don't like the idea of a function existing for a type that logically shouldn't be needed.

The idea of having `NodeType()` could work, but again this would be a meaningful function (as opposed to just having the unique function to define type membership). I'm not a big fan of this either, specifically around:

> Especially when you add a new non-leaf type and have to update everywhere that does recursion (although it kind of makes sense since how you do recursion changes on type)

I'm of the thinking that if a new nonterminal is introduced, the parser should absolutely explicitly deal with its processing at every point where nonterminals are processed. The converse, having an enum and checking with if-statements, makes it easy to forget to update parts of the logic. For languages with pattern matching, the fact that the compiler warns about non-exhaustive matching (to me) is an absolute win. We can't really do exhaustive checking in Go (where is that vet plugin @keegancsmith :P), but switching on the type, versus accessing a member function, assuages some of my concern these. In general, having an additional function to tell us the type feels redundant since we can type switch. The latter may be more verbose, and I'm  generally OK with that because the switch can also be exported to a non-member utility functions `IsX` or `NodeType()` for convenience, if that's really needed in some context. Using type switches also means we could use @keegancsmith's vet plugin if that happens.

So: I'll define a private member function for `Node` types but not use an additional mechanism like `NodeType()` or `isParam()`/`isOp()`, unless I've overlooked something in my discussion and there is a better motivation.